Hypernyms should always return a set

mboratko opened the pull request:
Previously, the `hypernyms` function was only returning a single synset, however there are some synsets with multiple hypernyms, eg. "lactate".

These changes make it so that `hypernyms` work the same as the other functions (`antonyms` and `hyponyms`), and abstracts the `expanded_hypernym` code to a general breadth-first-search `expanded_relation` function, which allows for `expanded_hyponyms` as well.

mboratko committed:
Hypernyms should always return a set

mboratko committed:
Abstracted expanded relation code, made hypernyms work with relation abstraction as a set

mboratko commented:
The tests failed because of the change in the return type (`relation` returning a set instead of an array), however I would argue that this is the correct return type in principle, so the downstream functions should be changed to expect a set as output.

oxinabox commented on src/operations.jl:
```diff
@@ -1,33 +1,33 @@
+        node = pop!(queue)
+        push!(all_related, node)
+        local_relatives = setdiff(relation(db, node, pointer_sym), all_related)
+        delete!(local_relatives, synset)
```
if you inititalized `all_related` to contain `synset` at the start,
then you would only need to delete it once at the end.

More broadly, though perhaps a it might be better to do

```
function expanded_relation(db::DB, synset::Synset, pointer_sym)
    all_related = Set{Synset}(synset)
    queue = relation(db, synset, pointer_sym)

    while !isempty(queue)
        node = pop!(queue)
        node in all_related && continue  # done this one already 
        push!(all_related, node)

        union!(queue, relation(db, node, pointer_sym))
    end

    delete!(all_related, synset)
    return all_related
end
```

since this avoids allocating a new set in `setdiff`

oxinabox commented:
Thanks.
I really like the `expanded_relation` and hyponym, support.

I am a little 🤷  to the change of everything to be `Sets`.
Because sets are generally slower than short vectors for all set operations in practice.
Because checking a vector of ~5 elements can be done in less time than most hash operations.
And wordnet relations are generally very small sets.

oxinabox commented:
@mboratko how is this going?

mboratko committed:
Use vector for direct relations

Vectors are faster than using sets, and WordNet already establishes uniqueness of single-relation expansion.

mboratko committed:
Use no-cycles assumption of WordNet to speed up expanded_relation

mboratko committed:
Test for multiple hypernym case added

oxinabox commented on src/operations.jl:
```diff
@@ -9,25 +9,26 @@ relation(db::DB, synset::Synset, pointer_sym) = map(
-antonyms(db::DB, synset::Synset)  = relation(db, synset, ANTONYM)
-hyponyms(db::DB, synset::Synset)  = relation(db, synset, HYPONYM)
+function expanded_relation(db::DB, synset::Synset, pointer_sym)
+    all_related = Set{Synset}([synset])
```
```suggestion
    # Inititally include self, so as to avoid cycles,
    # but we will delete it at the end 
    all_related = Set{Synset}([synset])
```

oxinabox committed:
add note about why including self

oxinabox commented:
LGTM, Thanks!
nice first contribution
